lib/instrumentation/trace-context.js

```diff
@@ -0,0 +1,84 @@
+'use strict'
+
+const crypto = require('crypto')
+
+const SIZES = {
+  version: 1,
+  traceId: 16,
+  id: 8,
+  flags: 1,
+
+  // Aggregate sizes
+  ids: 24, // traceId + id
+  all: 26
+}
+
+const OFFSETS = {
+  version: 0,
+  traceId: SIZES.version,
+  id: SIZES.version + SIZES.traceId,
+  flags: SIZES.version + SIZES.ids
+}
+
+function defineLazyProp (obj, prop, fn) {
+  Object.defineProperty(obj, prop, {
+    configurable: true,
+    enumerable: true,
+    get () {
+      const value = fn()
+      Object.defineProperty(obj, prop, {
+        enumerable: true,
+        value
+      })
+      return value
+    }
+  })
+
+  return obj
+}
+
+function hexSliceFn (buffer, offset, length) {
+  return () => buffer.slice(offset, length).toString('hex')
+}
+
+const bufferSymbol = Symbol('trace-context-buffer')
+
+class TraceContext {
+  constructor (buffer) {
+    this[bufferSymbol] = buffer
+    defineLazyProp(this, 'version', hexSliceFn(buffer, OFFSETS.version, OFFSETS.traceId))
+    defineLazyProp(this, 'traceId', hexSliceFn(buffer, OFFSETS.traceId, OFFSETS.id))
+    defineLazyProp(this, 'id', hexSliceFn(buffer, OFFSETS.id, OFFSETS.flags))
+    defineLazyProp(this, 'flags', hexSliceFn(buffer, OFFSETS.flags))
+  }
+
+  static create (sampled = false) {
+    const buffer = Buffer.alloc(SIZES.all)
+    crypto.randomFillSync(buffer, OFFSETS.traceId, SIZES.ids)
+    if (sampled) buffer[OFFSETS.flags] = 0x01
+    return new TraceContext(buffer)
+  }
+
+  static fromString (header) {
+    return new TraceContext(Buffer.from(header.replace(/-/g, ''), 'hex'))
```

Comment:
We should probably add a comment about how we're currently not following all the rules about validating the version byte, the length of the header etc as specified here: https://w3c.github.io/distributed-tracing/report-trace-context.html#versioning-of-traceparent

It's fine now as we're using our own custom header name - i.e. it's not a blocker - so just a comment would suffice.